cmd/kubeadm/app/phases/upgrade/upgrade.go

```diff
@@ -0,0 +1,152 @@
+			fmt.Printf("\t- OPTION 1: Run 'kubeadm config upload from-flags' and specify the same CLI arguments you passed to 'kubeadm init' when you created your control-plane.\n")
+			fmt.Printf("\t- OPTION 2: Run 'kubeadm config upload from-file' and specify the same config file you passed to 'kubeadm init' when you created your control-plane.\n")
+			fmt.Printf("\t- OPTION 3: Pass a config file to 'kubeadm upgrade' using the --config flag.\n")
+			fmt.Println("")
+			err = errors.Errorf("the ConfigMap %q in the %s namespace used for getting configuration information was not found", constants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+		}
+		return nil, nil, nil, errors.Wrap(err, "[upgrade/config] FATAL")
+	}
+
+	return client, cfg, NewOfflineVersionGetter(NewKubeVersionGetter(client), newK8sVersion), nil
+}
+
+// EnforceRequirements verifies that it's okay to upgrade and then returns the variables needed for the rest of the procedure
+func EnforceRequirements(client clientset.Interface, cfg *kubeadmapi.InitConfiguration, ignorePreflightErrorsSet sets.String, featureGatesString string, dryRun bool) error {
+	// Check if the cluster is self-hosted
+	if IsControlPlaneSelfHosted(client) {
+		return errors.New("cannot upgrade a self-hosted control plane")
+	}
+
+	// Fetch the configuration from a file or ConfigMap and validate it
+	fmt.Println("[upgrade/config] Making sure the configuration is correct:")
+
+	var err error
+	// If features gates are passed to the command line, use it (otherwise use featureGates from configuration)
+	if featureGatesString != "" {
+		cfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, featureGatesString)
+		if err != nil {
+			return errors.Wrap(err, "[upgrade/config] FATAL")
+		}
+	}
+
+	// Check if feature gate flags used in the cluster are consistent with the set of features currently supported by kubeadm
+	if msg := features.CheckDeprecatedFlags(&features.InitFeatureGates, cfg.FeatureGates); len(msg) > 0 {
+		for _, m := range msg {
+			fmt.Printf("[upgrade/config] %s\n", m)
+		}
+		return errors.New("[upgrade/config] FATAL. Unable to upgrade a cluster using deprecated feature-gate flags. Please see the release notes")
+	}
+
+	// Ensure the user is root
+	klog.V(1).Info("running preflight checks")
+	if err := RunPreflightChecks(client, ignorePreflightErrorsSet, &cfg.ClusterConfiguration); err != nil {
+		return err
+	}
+
+	// Run healthchecks against the cluster
+	if err := CheckClusterHealth(client, &cfg.ClusterConfiguration, ignorePreflightErrorsSet); err != nil {
+		return errors.Wrap(err, "[upgrade/health] FATAL")
+	}
+
+	// Use a real version getter interface that queries the API server, the kubeadm client and the Kubernetes CI system for latest versions
+	return nil
+}
+
+// GetClient gets a real or fake client depending on whether the user is dry-running or not
+func GetClient(file string, dryRun bool) (clientset.Interface, error) {
+	if dryRun {
```

Comment:
can return early here and avoid indenting the bigger branch for `if dryRun...`
```
if !dryRun {
    return kubeconfigutil.ClientSetFromFile(file)
}
...
```

Reply:
ok. my original goal was only to change what I needed. I will make this secondary changes to improve flow in their own commit.